### ISSUE TITLE
feat(router): swap / burn amount more than available token in pair

### ISSUE THREAD

**Issue by xenide** [vexchange/v3-core#98]
Users might want to swap an amount for which the amountOut is more than what the pair has (the rest being managed by the asset manager). Thus we need to introduce logic to pull some assets back from the asset manager in order to facilitate this swap. 

Note: we need to implement this in the router rather than the pair or the asset manager as we do not want to complicate the interface of the pair and asset manager. 

## Cases to take care of 

- [ ] output amount for swap is more than reserves 
- [ ] redeem amount during burn is more than reserves

**Comment by xenide** [comment]
So to implement this with the router, here's what I propose: 

For the router, in swap functions (both exact in and out), and the `removeLiquidity` function, it checks if the output amount is greater than its actual holdings in the pair (reserves - managed), such that the ERC20 transfer does not fail. 
- If so, router calls `assetManager.returnAsset()` to get the asset back from the 3rd party in order to facilitate this swap. If there are enough holdings to cover this trade, do nothing. 

For the router, additional variables:
```solidity
IAssetManager assetManager;
```

For the asset manager,

```solidity
// actually can just use the existing adjustManagement function
// but have to change the permissions to owner + router 
function returnAsset(address aPair, address aToken, uint256 aSufficientAmt) {
   // retrieve asset from AAVE 
   // transfer back to pair 
   // update managed amount
}

// to control who can call returnAsset
// could even just be an address instead of a mapping
mapping (address => bool) whitelistedRouters
```

When writing this, I realized something. Does `Pair::adjustManagement` need to call `_update` at the end? The reserves would still be the same cuz the ERC20 balance just becomes the managed amount.

**Comment by OliverNChalk** [comment]
What if the AM does not return the assets when requested?

**Comment by xenide** [comment]
> What if the AM does not return the assets when requested?

Then the swap would just fail because there is not enough token to give out. Specifically, it will fail when the `token.transfer()` function is called as there is insufficient balance

**Comment by OliverNChalk** [comment]
Would it be better for returnAsset to revert if the AM refuses to? That way front-end can tell if it is possible to do this swap route

**Comment by xenide** [comment]
In the current impl, `adjustManagement`/`returnAsset` would revert by itself, from AAVE's error. 

I just looked at the code again and it seems that if the pool were inactive or paused, the tx would revert. I wasn't aware of this possibility when I first wrote it. So, do you think we should handle the failure case gracefully or it can just revert by itself? 

I'd say we don't have a HARD need to fail gracefully in this case since the entire tx would just revert. But if we wanted, when the asset mgr automatically / manually moves assets and it reverts, we can put the assets back into the pair and make sure the accounting of assets is correct. What do you think?

**Comment by OliverNChalk** [comment]
I do not follow, when you say fail gracefully do you mean catch the aave error and return a different error? Or somehow avoid a revert altogether?

**Comment by xenide** [comment]
1. UniV3 style callbacks.
 a. Increases code complexity in router as it will need to treat the normal vs rebalance swaps quite differently. (cannot just do the rebalances at the start of execution because the callback will force the router to return to the AM between swaps).
2. Gatekeep the overRebalance function so only the router can call (can be a mapping to support multiple routers).
 a. This prevents 3rd party routers (1inch etc) and MEV bots from being able to use this function.


Note, we can implement both if we want the best of both worlds.

**Comment by xenide** [comment]
After marinating for some time, I'm leaning towards option 2. 

Option 1 seems a little complicated implementation wise. Unless you want to explain to me further how this is the better solution? 

For Option 2, 3rd party routers bots can still swap the amount available in the pair. And they can know the limit for that by calling 2 functions (`balanceOf` and `token0/1Managed`). It's not blocking them out completely. 

If we do option 2, decisions we need to make: 
1. during the return of assets, do we want to pull more than we have to? 
   - are we okay with having 0 tokens in the pair and all remaining tokens in the asset manager? 
   - or do we want to pull more according to that ratio, to make sure if there's a swap in the near future, there will be some tokens to cover it
   - my preference now is leaning towards not pulling more than we have to. i.e. lazy loading / allocation principle. If subsequent txs need more assets returned, the asset manager + router can do the same thing

**Comment by OliverNChalk** [comment]
> are we okay with having 0 tokens in the pair and all remaining tokens in the asset manager?

During the return of the assets we mandate that the final state is optimal, where optimal is defined by the AM. So it cannot result in zero tokens in the pair.

> or do we want to pull more according to that ratio, to make sure if there's a swap in the near future, there will be some tokens to cover it

See above

> my preference now is leaning towards not pulling more than we have to. i.e. lazy loading / allocation principle. If subsequent txs need more assets returned, the asset manager + router can do the same thing

This results in excessive gas usage, I propose leaving the pool in an optimal state (why not, it costs the same amount of gas).


Follow up question, why not both 1 and 2, technical complexity? That said, given this is implmeneted in the AM, we can always do 2 and then do 1 later if there is demand.

**Comment by xenide** [comment]
Could you outline a mock of methods and some pseudocode for option 1? I'm having a lil trouble visualizing in my head. 

Also, what is the marginal benefit of option 1 if we're going to implement option 2?

**Comment by OliverNChalk** [comment]
```solidity
function swapWithCapital(aPair, aToken, aAmount) external lock {
    aToken.transfer(aPair, aAmount);
    IAmCallback(msg.sender).callAfterCapital();
    require(aToken.balanceOf(aPair) == _optimalAmount(aPair, aToken), "Imbalanced Pool");
}
```

**Comment by xenide** [comment]
to be very clear here, this function `swapWithCapital` will be in the router? And the callback `callAfterCapital` will be implemented by the asset manager, which returns assets to the pair? 

Then who is `msg.sender`? 

So the marginal benefit is that external routers do not have to worry about the managed amounts vs actual amounts? That the swaps will always succeed?

**Comment by OliverNChalk** [comment]
No, this function is in the AM

**Comment by xenide** [comment]
Now that I understand the intricacies of things, I'd like to (re)propose the third option, which is letting the pair calculate and decide how much assets to pull back. 

It seems to me that this approach has the following advantages: 

- encapsulates the logic where it is supposed to be. `swap` and `burn` would be the ones to take care of pulling additional assets back when the need arises, no other entities need to care about this logically. 
- NO ONE would not have to worry about any AM issues. If there's reserves, it can be swapped

Yes, we'd need to add an additional function in the AM to facilitate this, like a `returnAssetToMe` which the `Pair`s call. We'd have to force this `AaveManager` and future managers to implement this method yes. But this seems like a reasonable trade off to me, compared to the overlaps in scope and complexity we're going to have for option 1 and 2. 

Remind me again why this solution is so bad?

**Comment by OliverNChalk** [comment]
How much extra gas cost for swaps that don't use this feature?

**Comment by xenide** [comment]
so the flow goes like this: 

```solidity
// checks to see if output required is greater than what is in the pair now
if (reserveOut - tokenOutManaged < outputAmt) {
   assetManager.returnAssetToMe(requiredAmount);
}

// optimistically transfers tokens 
```

At first I thought we had to do an `IERC20.balanceOf(this)`, but turns out we can just calculate how much there is in the pair by subtracting the managed amount from reserves. 

Therefore, it is an additional `sload` to read the managed amount for ALL swaps. For swaps that require assets returned, an additional call to the asset manager.

**Comment by OliverNChalk** [comment]
So an extra `2100 gas` for all swaps if we use this approach. What about a `tryTransfer()` approach that only does the check if the transfer reverts, would that be more gas efficient?

**Comment by xenide** [comment]
With my unskillful means, 

```solidity
    function testTryTransfer() public
    {
        _tokenA.mint(address(this), 100);
        try _tokenA.transfer(_bob, 100) {

        }
        catch {

        }
    }

    function testOnlyTransfer() public
    {
        _tokenA.mint(address(this), 100);
        _tokenA.transfer(_bob, 100);
    }
```

first one costs `45778` while the second costs `45710`

When I simulated a cold `SLOAD` of an `uint256` on the second test case, the gas cost went up to `47642 `. So it seems like using `try/catch` is the more gas efficient option.

However, I notice that our inherited implementation uses `_safeTransfer`, which checks the `bool` of the returns and then length of the returned data. So if we were to implement this on top of the existing `_safeTransfer`, I guess we might not need to use `try/catch` but just check return values of the low level calls? This impl seems to cost even fewer gas (`42908`) than the `try/catch` in my experiments. 

What do you think?

**Comment by OliverNChalk** [comment]
64512 - 45710 = 18802
SLOAD costs 2100

I would not put much faith in the numbers you are looking at, why is the cost of an SLOAD 18k?

**Comment by xenide** [comment]
Finally figured out. okay I was a retard. I think that's due to it writing a non-zero value into storage. The amount I was transferring was not 100. 

So actually it costs 47642 for the `SLOAD` scenario. In any case, `try/catch` seems more gas efficient than `SLOAD`, and `.call()` is the cheapest it seems. 

But due to safety reasons (such as some ERC20s not returning `true` for a successful `transfer`), should we build on top of `.call()` and then checking those results?

**Comment by OliverNChalk** [comment]
Open a PR with your best impl then lets talk about real code rather than abstract approachs - its a little hard for me to visualize. Should only be a few lines so much easier to just write them then talk

**Comment by xenide** [comment]
closing as this is addressed is #96
